Documentation updates for v12 (#4880)
cypress-io/cypress-documentation: content/guides/references/trade-offs.md

- There will never be support for [multiple browser tabs](#Multiple-tabs).
- You cannot use Cypress to drive
  [two browsers at the same time](#Multiple-browsers-open-at-the-same-time).
- Each test is bound to a [single superdomain](#Navigation-Rules). Cross-origin

>   navigation inside each test is permitted with the use of the

Reviewer: ```suggestion
  navigation inside tests can be enabled by using the
```

Author: addressed in c71644c13

This avoids ever needing a second open browser, but still gives you an
end-to-end test that provides 100% confidence that the two clients can
communicate with each other.

### Navigation Rules

> By itself, Cypress limits each test to only visiting domains that are considered
> to be same superdomain to one another. If a navigation occurs that does not meet

Reviewer: ```suggestion
Cypress limits each test to visiting domains that share the
same superdomain. If a navigation occurs that does not meet
```

Author: addressed in c71644c13

the same superdomain rule, the [`cy.origin`](/api/commands/origin) command must
be used to execute assertions inside the newly navigated origin.

> But what is same superdomain? It is actually very similar to that of same

Reviewer: I think the rest of this section should be moved to a Cross-Origin Testing guide and linked to.

origin! Two URLs have the same origin if the protocol, port (if specified), and
host are the same for both. Cypress automatically handles hosts of the same
superdomain by injecting the
[`document.domain`](https://developer.mozilla.org/en-US/docs/Web/API/Document/domain)
property into the visited `text/html` pages. This is why navigations without the

> use of the [`cy.origin`](/api/commands/origin) command are solely scope to same

Reviewer: ```suggestion
use of the [`cy.origin`](/api/commands/origin) command are solely scope to the same
```

Author: addressed in c71644c13

superdomain.

Given the URLs below, all have the same superdomain compared to
`https://www.cypress.io`.

- `https://cypress.io`
- `https://docs.cypress.io`
- `https://example.cypress.io/commands/querying`

The URLs below, however, will have different superdomains/origins compared to
`https://www.cypress.io`.

- `http://www.cypress.io` (Different protocol)
- `https://docs.cypress.io:81` (Different port)
- `https://www.auth0.com/` (Different host of different superdomain)

The `http://localhost` URLs differ if their ports are different. For example,
the `http://localhost:3000` URL is considered to be a different origin from the
`http://localhost:8080` URL.

The rules are:

- <Icon name="exclamation-triangle" color="red"></Icon> You **cannot**
  [visit](/api/commands/visit) two domains of different superdomains in the same

>   test and continue to run assertions without the use of the

Reviewer: ```suggestion
  test and continue to interact with the page without the use of the
```

Author: addressed in c71644c13

#### Other workarounds

There are other ways of testing the interaction between 2 superdomains. Because
the browser has a natural security barrier called `origin policy` this means
that state like `localStorage`, `cookies`, `service workers` and many other APIs
are not shared between them anyways. Because of this, the Cypress APIs that
interact with `localStorage`, `sessionStorage`, and `cookies` are able to take a
domain option in order to interact with the appropriate domain.

As a best practice, you should not visit or interact with a 3rd party service
not under your control. However, there are exceptions! If your organization uses
Single Sign On (SSO) or OAuth then you might involve a 3rd party service other
than your superdomain, which can be safely tested with
[`cy.origin`](/api/commands/origin).

We've written several other guides specifically about handling this situation.

- [Best Practices: Visiting external sites](/guides/references/best-practices#Visiting-external-sites)
- [Web Security: Common Workarounds](/guides/guides/web-security#Common-Workarounds)
- [Recipes: Logging In - Single Sign On](/examples/examples/recipes#Logging-In)

> - LINK AUTHENTICATION GUIDES HERE

Author: addressed in c71644c13

- `https://cypress.io`
- `https://docs.cypress.io`
- `https://example.cypress.io/commands/querying`

The URLs below, however, will have different superdomains/origins compared to
`https://www.cypress.io`.

- `http://www.cypress.io` (Different protocol)
- `https://docs.cypress.io:81` (Different port)
- `https://www.auth0.com/` (Different host of different superdomain)

The `http://localhost` URLs differ if their ports are different. For example,
the `http://localhost:3000` URL is considered to be a different origin from the
`http://localhost:8080` URL.

The rules are:

- <Icon name="exclamation-triangle" color="red"></Icon> You **cannot**
  [visit](/api/commands/visit) two domains of different superdomains in the same
  test and continue to run assertions without the use of the
  [`cy.origin`](/api/commands/origin) command.
- <Icon name="check-circle" color="green"></Icon> You **can**
  [visit](/api/commands/visit) two or more domains of different origin in
  **different** tests without needing [`cy.origin`](/api/commands/origin).

We understand this is a bit complicated to understand, so we have built a nifty
chart to help clarify the differences!

#### Parts of a URL

> ```

Author: @chrisbreiding added the parts of a url chart here that we can link through our other docs. This area felt appropriate but lmk what you think.

it('navigates', () => {
  cy.visit('https://www.cypress.io')
  cy.visit('https://stackoverflow.com')
  cy.origin('https://stackoverflow.com', () => {
    cy.get('selector') // yup all good
  })
})
```

```javascript
it('navigates', () => {
  cy.visit('https://www.cypress.io')
})

// split visiting different origin in another test
it('navigates to new origin', () => {
  cy.visit('https://stackoverflow.com')
  cy.get('selector') // yup all good
})
```

This limitation exists because Cypress switches to the domain under each
specific test when it runs. For more information on this, please see our Web
Security page regarding
[Different superdomain per test requires cy.origin command](/guides/guides/web-security#Different-superdomain-per-test-requires-cy-origin-command).

#### Other workarounds

There are other ways of testing the interaction between 2 superdomains. Because
the browser has a natural security barrier called `origin policy` this means
that state like `localStorage`, `cookies`, `service workers` and many other APIs
are not shared between them anyways. Because of this, the Cypress APIs that
interact with `localStorage`, `sessionStorage`, and `cookies` are able to take a

> domain option in order to interact with the appropriate domain.

Reviewer: The local/sessionStorage will be ready, but they _do not_ accept a domain option. They are the getAll/clearAll variety. So the (currently implemented/planned) options will be:

- get/set/clear cookies(s): use current origin by default, but accept domain option
- get/clear all local/session storage: get or clear them all
- get/clear all cookies: get or clear them all (not guaranteed to make the 12.0.0 release)
